### PR TITLE
Support AWQ models

### PR DESCRIPTION
**Pull request by mvafin**
# What does this PR do?

Add support for AWQ models after this support was added by OpenVINO in 2024.6 (https://github.com/openvinotoolkit/openvino/pull/27859)
Patching for 16bit can be also used for GPTQ and AWQ models to support FP16/BF16 regions in the model.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Review by eaidova — APPROVED** [review]

**Comment by AlexKoff88** [comment]
@mvafin, can we have a test for this, e.g. with "hf-internal-testing/Mixtral-tiny-AWQ" or any other dummy model?

**Comment by IlyasMoutawwakil** [comment]
> @mvafin, can we have a test for this, e.g. with "hf-internal-testing/Mixtral-tiny-AWQ" or any other dummy model?

yes a test would be great

**Comment by mvafin** [comment]
@AlexKoff88 @IlyasMoutawwakil Test added

**Comment by AlexKoff88** [comment]
@mvafin, some of the tests failed:
```
=========================== short test summary info ============================
FAILED tests/openvino/test_modeling.py::OVModelForCausalLMIntegrationTest::test_compare_to_transformers_18_mixtral_awq - ValueError: `.float()` is not supported for quantized model. Please use the model as it is, since the model has already been casted to the correct `dtype`.
FAILED tests/openvino/test_modeling.py::OVModelForCausalLMIntegrationTest::test_compare_to_transformers_23_qwen - AssertionError: Torch not compiled with CUDA enabled
FAILED tests/openvino/test_modeling.py::OVModelForSeq2SeqLMIntegrationTest::test_compare_with_and_without_past_key_values - AssertionError: False is not true : With pkv latency: 227.547 ms, without pkv latency: 240.161 ms, speedup: 1.055
FAILED tests/openvino/test_modeling.py::OVModelForSTFeatureExtractionIntegrationTest::test_compare_to_transformers_0_st_bert - AssertionError: Torch not compiled with CUDA enabled
FAILED tests/openvino/test_modeling.py::OVModelForSTFeatureExtractionIntegrationTest::test_compare_to_transformers_1_st_mpnet - AssertionError: Torch not compiled with CUDA enabled
```

Can you please check on your side?

**Comment by mvafin** [comment]
> @mvafin, some of the tests failed:
> 
> ```
> =========================== short test summary info ============================
> FAILED tests/openvino/test_modeling.py::OVModelForCausalLMIntegrationTest::test_compare_to_transformers_18_mixtral_awq - ValueError: `.float()` is not supported for quantized model. Please use the model as it is, since the model has already been casted to the correct `dtype`.
> FAILED tests/openvino/test_modeling.py::OVModelForCausalLMIntegrationTest::test_compare_to_transformers_23_qwen - AssertionError: Torch not compiled with CUDA enabled
> FAILED tests/openvino/test_modeling.py::OVModelForSeq2SeqLMIntegrationTest::test_compare_with_and_without_past_key_values - AssertionError: False is not true : With pkv latency: 227.547 ms, without pkv latency: 240.161 ms, speedup: 1.055
> FAILED tests/openvino/test_modeling.py::OVModelForSTFeatureExtractionIntegrationTest::test_compare_to_transformers_0_st_bert - AssertionError: Torch not compiled with CUDA enabled
> FAILED tests/openvino/test_modeling.py::OVModelForSTFeatureExtractionIntegrationTest::test_compare_to_transformers_1_st_mpnet - AssertionError: Torch not compiled with CUDA enabled
> ```
> 
> Can you please check on your side?

Fixed

**Comment by IlyasMoutawwakil** [comment]
still getting awq cuda errors

**Comment by mvafin** [comment]
> still getting awq cuda errors

That might be because optimum is not tested with 2024.6 version of openvino

**Review by AlexKoff88 — APPROVED** [review]

File: .github/workflows/test_openvino.yml
Comment (IlyasMoutawwakil) [inline .github/workflows/test_openvino.yml:57]: these are not valid extra urls for auto-gptq and awq

**Comment by IlyasMoutawwakil** [comment]
The PR in its current form seems to degrade the user experience greatly. My understanding is that it suggests we pass the responsibility of patching autoawq and autogptq to the user ? like writing their own `mock_torch_cuda_is_available` and `patch_awq_for_inference` to use autogptq/autoawq. Please correct me if I'm wrong.

File: .github/workflows/test_openvino.yml
Comment (eaidova) [reply .github/workflows/test_openvino.yml:57]: this is neede for preventing reinstalling torch with cuda during installing third-party, packages themselves should be installed from regular source, torch-dependent libs (the difference from --index-url and --extra-index-url that first redefine source index completely, the second one parameter used for usage index URL as additional source if library present in that source) will be tried to install from torch cpu url

**Comment by eaidova** [comment]
> The PR in its current form seems to degrade the user experience greatly. My understanding is that it suggests we pass the responsibility of patching autoawq and autogptq to the user ? like writing their own `mock_torch_cuda_is_available` and `patch_awq_for_inference` to use autogptq/autoawq. Please correct me if I'm wrong.

these manual patches requires only for running original torch model without cuda in test environment, general flow for model conversion does not requires these patches, so no any impact on optimum-intel user experience as running original model in torch is out of our responsibility

**Comment by IlyasMoutawwakil** [comment]
> these manual patches requires only for running original torch model without cuda in test environment

Thanks for the explanation ! I thought it's also required for the quantization process on cpu but I see that patch is still in its place.

File: optimum/exporters/openvino/__main__.py
Comment (IlyasMoutawwakil) [inline optimum/exporters/openvino/__main__.py:253]: why patch `auto-gptq` lib when method is `awq` ?

File: tests/openvino/test_modeling.py
Comment (eaidova) [reply tests/openvino/test_modeling.py:1001]: slow test with beam search testing and generated sequences in the same test are equal. logits can be represented in denormalized format and it is hard to say which threshold is meaningful as we do not know range of values. If I apply softmax on output I see 1-e5 difference between torch and ov result, argsort of both tensors shows small permutations after 1000 tokens in logits probs (it is just misplacing 1027 and 1028 tokens), so I believe it is enough accurate to avoid impact on generated strings

Comment (eaidova) [reply tests/openvino/test_modeling.py]: added skipping logits comparision for quantized models

**Review by IlyasMoutawwakil — APPROVED** [review]
LGTM ! Merging approved as long as tests are passing ! Thanks @eaidova for following up on this.

**Comment by eaidova** [comment]
@IlyasMoutawwakil  looks like all tests passing now, let me fix code style after rebasing and I think everything is done for this PR

**Comment by AlexKoff88** [comment]
Merging as we have few approvals.

**Pull request merged**
